### ISSUE TITLE
Reducing memory usage during cnvkit.py segment for WGS samples

### ISSUE THREAD

**Issue by chapmanb** [etal/cnvkit#346]
Eric;
Recently we've hit some issues with cnvkit.py segment using a large amount of memory (>60Gb) and exceeding memory requirements on a SLURM setup. These are relatively low coverage WGS samples (~20Gb BAMs) and when digging into the and reads they seemed to fail in chromosomes with high coverage spikes (>2000x coverage) due to repetitive elements.

I was wondering if you had any general tips/tricks for reducing memory usage or generally making how we run this in bcbio better. Here is how we build up the command line in bcbio:

https://github.com/bcbio/bcbio-nextgen/blob/085acd04b2511936b3f9ad10e67f3476e59be3e5/bcbio/structural/cnvkit.py#L280

For WGS on these runs we have 16 cores (`-p 16`) and lower the segmentation threshold (`--threshold 0.00001`) but don't use VCF inputs. This was a tumor-only sample, so we added `--drop-low-coverage`. I was able to work around/mitigate the issue by using less cores (8 instead of 16) but wanted to check if there were more specific things we could try to improve how we use memory. Thanks for any tips or tricks.

**Comment by etal** [comment]
I think the bottleneck is in the `segment` command's internal post-processing after CBS is run. If there are lots of segments, and/or lots of bins, the intersection method blows up. I think the solution is to change the way the intersection method works here, reducing the number of intermediate objects. (It's a priority.)

You've already lowered the segmentation threshold and dropped low-coverage bins. Using fewer cores is the quick workaround. You could also try being more aggressive in masking out problematic genomic regions with `access -x` -- based on your benchmarks you could basically use a BED of all the regions with frequent false-positive calls as another `-x` input to be used with all unpaired CNVkit runs.

**Comment by chapmanb** [comment]
Eric -- thanks so much for looking into a better implementation and for the tips on working around in the short term. The false positive exclusion is a great idea, I need to work on building up a collection of these regions to use for this. I'd definitely be happy to try out the new intersection method when you think it is ready and will work around it with core adjustment in the short term. Thanks again for the help.

**Comment by etal** [comment]
Some speedups to `segment` are in place now, and I think I can do more in `segmetrics` to streamline the bootstrap resampling procedure.

In the CNVkit repo, the example files `access-5k-mappable.{hg19,grch37}.bed` use the Ensembl DAC and Duke low-mappability regions for exclusion from the corresponding reference genomes. I took those files from UCSC goldenPath.

Which version of pandas is bcbio using now? I think there are some performance improvements in recent releases, and I saw that the relevant conda channels have built packages atop newer versions of numpy, which had limited pandas to version 0.18.1 earlier.
